### PR TITLE
feat(nns): Support upgrading root canister through install_code

### PR DESCRIPTION
**Pull request by jasonz-dfinity**
# Why

Previously, `install_code` proposal action is added without supporting upgrading root. This PR supports it.

# Changes

We support creating a proposal with action `install_code` while specifying NNS Root as the target canister, and the proposal execution code will call the correct `upgrade_root` endpoint correctly (unlike upgrading other canisters which calls root). Note that `install` or `reinstall` modes aren't supported. It might not be obvious why `reinstall` isn't supported - HardResetRootToVersion supports "resetting" root, but the reset is implemented as an `uninstall_code` followed by `install_code` (with `Install` mode). This behavior seems to exceed what can be reasonably thought of as the `install_code` proposal action. Therefore, `HardResetRootToVersion` will remain as an NnsFunction and will later be switched to `ProtocolCanisterManagement` topic.

### CODE REVIEW

File: rs/nns/governance/src/proposals/install_code.rs [hunk 1]
```diff
@@ -102,21 +106,21 @@ impl InstallCode {
...
-        if canister_id == ROOT_CANISTER_ID {
-            // TODO(NNS1-3190): support changing root canister
-            return Err(Self::invalid_proposal_error(
-                "InstallCode for root canister is not supported yet",
-            ));
-        }
-        Ok((ROOT_CANISTER_ID, "change_nns_canister"))
+    fn payload_to_upgrade_root(&self) -> Result<Vec<u8>, GovernanceError> {
+        let stop_upgrade_start = !self.skip_stopping_before_installing.unwrap_or(false);
+        let wasm_module = vec![];
```
Comment (daniel-wong-dfinity-org) [inline rs/nns/governance/src/proposals/install_code.rs:112]: Uh. Where code?? I assume you meant to use self.wasm_module, but forgot?? How can the assert on line 378 be passing??

File: rs/nns/governance/src/proposals/install_code.rs [hunk 2]
```diff
@@ -141,6 +145,37 @@ impl CallCanister for InstallCode {
...
+        }
+
+        let install_mode = self.valid_install_mode()?;
+        match install_mode {
+            RootCanisterInstallMode::Install | RootCanisterInstallMode::Reinstall => {
+                Err(Self::invalid_proposal_error(&format!(
+                    "InstallCode mode {:?} is not supported for root canister, consider using \
+                     HardResetNnsRootToVersion proposal instead",
+                    install_mode
+                )))
```
Comment (daniel-wong-dfinity-org) [inline rs/nns/governance/src/proposals/install_code.rs:162]: No change.

Would it be feasible to add support for this later? I read the PR description, and I understand that not supporting this is intentional, and is the plan going forward, but I'm not really understanding the why. Maybe, I need to re-read the description...

**Review by max-dfinity — APPROVED** [review]

File: rs/nns/governance/src/proposals/install_code.rs [hunk 3]
```diff
@@ -141,6 +145,37 @@ impl CallCanister for InstallCode {
     }
 }
 
+impl CallCanister for InstallCode {
+    fn canister_and_function(&self) -> Result<(CanisterId, &str), GovernanceError> {
```
Comment (daniel-wong-dfinity-org) [inline rs/nns/governance/src/proposals/install_code.rs:149]: Maybe add a triple slash comment that says something like

> Most canisters are upgraded indirectly via root. In such cases, we call root's change_nns_canister method. The exception is when root is to be upgraded. In that case, upgrades are instead done via lifeline's upgrade_root method.

File: rs/nns/governance/src/proposals/install_code.rs [hunk 4]
```diff
@@ -32,6 +33,15 @@ const PROTOCOL_CANISTER_IDS: [&CanisterId; 12] = [
     &EXCHANGE_RATE_CANISTER_ID,
 ];
 
+// The paylod type for the upgrade_root method in the lifeline canister. Should be kept in sync with
+// `rs/nns/handlers/lifeline/impl/lifeline.mo`.
```
Comment (daniel-wong-dfinity-org) [inline rs/nns/governance/src/proposals/install_code.rs:37]: NCR.

```suggestion
// When calling lifeline's upgrade_root method, this is the request. Keep this in sync with
// `rs/nns/handlers/lifeline/impl/lifeline.mo`.
```

File: rs/nns/governance/src/proposals/install_code.rs [hunk 5]
```diff
@@ -239,13 +274,28 @@ mod tests {
...
+                install_mode: Some(CanisterInstallMode::Install as i32),
                 ..valid_install_code.clone()
             }
             .validate(),
             Err(GovernanceError::new_with_message(
                 ErrorType::InvalidProposal,
-                "InstallCode proposal invalid because of InstallCode for root canister is not \
-                 supported yet"
+                "InstallCode proposal invalid because of InstallCode mode Install is not supported for root canister, \
+                consider using HardResetNnsRootToVersion proposal instead"
```
Comment (daniel-wong-dfinity-org) [inline rs/nns/governance/src/proposals/install_code.rs:284]: I would encourage you to not be so strict on the exact contents of the message. We should avoid rewording causing test failures. Admittedly, in practice, clients sometimes rely on the exact wording, but this kind of condones that.

[Here is an example of an alternative](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/nns/cmc/src/main.rs?L2542-2550).

File: rs/nns/governance/src/proposals/install_code.rs [hunk 6]
```diff
@@ -299,6 +349,38 @@ mod tests {
         );
     }
 
+    #[cfg(feature = "test")]
+    #[test]
+    fn test_upgrade_root_protocol_canister() {
+        let install_code = InstallCode {
+            canister_id: Some(ROOT_CANISTER_ID.get()),
+            wasm_module: Some(vec![1, 2, 3]),
```
Comment (daniel-wong-dfinity-org) [inline rs/nns/governance/src/proposals/install_code.rs:357]: NC: Future work?

We maybe should do some validation on this field. Specifically, there are ways to detect whether it is gzipped or a raw wasm.

File: rs/nns/governance/src/proposals/install_code.rs [hunk 7]
```diff
@@ -299,6 +349,38 @@ mod tests {
...
+            install_mode: Some(CanisterInstallMode::Upgrade as i32),
+            arg: Some(vec![4, 5, 6]),
```
Comment (daniel-wong-dfinity-org) [inline rs/nns/governance/src/proposals/install_code.rs:359]: NC: Future work?

We could maybe check that this is Candid-encoded. We don't necessary have to read the whole thing. I think Candid has some magic bytes at the beginning.

**Review by daniel-wong-dfinity-org — APPROVED** [review]

File: rs/nns/governance/src/proposals/install_code.rs (hunk 1 above)
Comment (jasonz-dfinity) [reply rs/nns/governance/src/proposals/install_code.rs:112]: Nice catch! I had the implementation working before, but I removed it to make sure that the state machine test fails without it (and it does), and forgot to put it back... Line 378 does also fail with this version. Fixed now.

File: rs/nns/governance/src/proposals/install_code.rs (hunk 2 above)
Comment (jasonz-dfinity) [reply rs/nns/governance/src/proposals/install_code.rs:162]: I don't see any problem with supporting it, but I just don't see any reason to do so.  Added some comments.

File: rs/nns/governance/src/proposals/install_code.rs (hunk 3 above)
Comment (jasonz-dfinity) [reply rs/nns/governance/src/proposals/install_code.rs:149]: Added comment to the if clause. IMHO comments on the method should describe its behavior (returns canister id and the function name for proposal execution), while the comment suggested here seems to be describing an implementation detail.

File: rs/nns/governance/src/proposals/install_code.rs (hunk 4 above)
Comment (jasonz-dfinity) [reply rs/nns/governance/src/proposals/install_code.rs:37]: Done.

File: rs/nns/governance/src/proposals/install_code.rs (hunk 7 above)
Comment (jasonz-dfinity) [reply rs/nns/governance/src/proposals/install_code.rs:359]: I think in most of the cases we can but not sure if we can say it's always the case. Ironically, NNS Governance init args is protobuf-encoded...

File: rs/nns/governance/src/proposals/install_code.rs (hunk 6 above)
Comment (jasonz-dfinity) [reply rs/nns/governance/src/proposals/install_code.rs:357]: Added a comment on the `validate()` to say we could do that. However, I doubt if we want to commit to that (so, not a TODO). I'm pretty sure the current NnsFunction does not validate it.

File: rs/nns/governance/src/proposals/install_code.rs (hunk 5 above)
Comment (jasonz-dfinity) [reply rs/nns/governance/src/proposals/install_code.rs:284]: Done.

**Pull request merged**